lib/elixir/lib/macro.ex

```diff
@@ -245,7 +245,33 @@ defmodule Macro do
   end
 
   @doc """
-  Pipes `expr` into the `call_args` at the given `position`.
+  Pipes the `expr` into the `call_args` at the given `position`.
+
+  `expr` is the AST of an expression. `call_args` must be the AST *of a call*,
+  otherwise this function will raise an error.
+
+  ## Examples
+
+      iex> ast = Macro.pipe(quote(do: div(10, 2)), quote(do: div(5)), 0)
+      {:div, [], [{:div, [context: MacroTest, import: Kernel], [10, 2]}, 5]}
+      iex> Code.eval_quoted(ast)
+      {1, []}
+
+  If attempted to pipe an expression into something other than a call,
+  the following error will be raised:
+
+      iex> Macro.pipe(10, 20, 0)
+      ** (ArgumentError) cannot pipe 10 into 20, can only pipe into local calls foo(), remote calls Foo.bar() or anonymous function calls foo.()
+
+  Even if the expression is piped into the AST, it doesn't
+  necessarily mean that the AST is valid and it might generate runtime issues,
+  for example:
+
+      iex> ast = Macro.pipe(10, {:div, [], [2, 5]}, 0)
+      {:div, [], [10, 2, 5]}
+      iex> Code.eval_quoted(ast)
+      ** (CompileError) nofile:1: undefined function div/3 (there is no such import)
+
```

Comment:
I am not the biggest fun of calling `Code.eval_quoted` in the tests. There is rarely a use case for calling `eval_quoted` and worry that if others see our examples, they will copy it. Thoughts @whatyouhide?

Reply:
Nice! thanks for the insightful comments.

Maybe we can just leave the explanation that even if its piped, it might not be valid. That sounds good?

Reply:
Yes. you can say something like "For example, you could pipe an argument to `div/2` effectively turning it into a call to `div/3`, which is a function that won't exist at runtime" or something along those lines.